pivot.py

```diff
@@ -0,0 +1,1723 @@
+            remove_field = False
+            if table_property:
+                tp = {}
+                for field in table_property.split('&'):
+                    if field:
+                        if field.split('=')[0] == 'hierarchy':
+                            tp[field.split('=')[0]] = int(field.split('=')[1])
+                        else:
+                            tp[field.split('=')[0]] = field.split('=')[1]
+                        if (field.split('=')[0] == 'name' and
+                                field.split('=')[1] == self.field):
+                            remove_field = True
+                if remove_field:
+                    hierarchy = tp['hierarchy']
+
+                if not remove_field:
+                    match tp['position']:
+                        case 'x':
+                            x_fields.append(tp)
+                        case 'y':
+                            y_fields.append(tp)
+                        case 'aggregation':
+                            aggregation_fields.append(tp)
+
+        # Sort each list using the "hierarchy" field
+        if x_fields:
+            if self.header == 'x':
+                for x_field in x_fields:
+                    if x_field['hierarchy'] > hierarchy:
+                        x_field['hierarchy'] -= 1
+            x_fields = sorted(x_fields, key=lambda x: x['hierarchy'])
+        if y_fields:
+            if self.header == 'y':
+                for y_field in y_fields:
+                    if y_field['hierarchy'] > hierarchy:
+                        y_field['hierarchy'] -= 1
+            y_fields = sorted(y_fields, key=lambda x: x['hierarchy'])
+        if aggregation_fields:
+            if self.header == 'y':
+                for aggregation_field in aggregation_fields:
+                    if aggregation_field['hierarchy'] > hierarchy:
+                        aggregation_field['hierarchy'] -= 1
+            aggregation_fields = sorted(aggregation_fields, key=lambda x: x['hierarchy'])
+
+        # Transform again to string
+        table_properties = ''
+        for field in x_fields:
+            table_properties += f'name={field["name"]}&position={field["position"]}&hierarchy={field["hierarchy"]}&__'
+        for field in y_fields:
+            table_properties += f'name={field["name"]}&position={field["position"]}&hierarchy={field["hierarchy"]}&__'
+        for field in aggregation_fields:
+            table_properties += f'name={field["name"]}&position={field["position"]}&aggregation={field["aggregation"]}&hierarchy={field["hierarchy"]}&__'
+
+        # In case we delete all the fields
+        if table_properties == '':
+            table_properties = 'null'
+        # Send the new table properties (redirect to index)
+        return redirect(Index(database_name=self.database_name, table_name=self.table_name, table_properties=table_properties, render=False).url())
+
+    def level_up_field(self):
```

Comment:
This and other methods have much of the code duplicated due to the lack of an abstraction for the information encoded in the URL.

Two methods are needed: `encode_url` and `decode_url` which specialize in the conversion of the URL into an easy to manipulate object. Probable an implementation of a class `URL` which includes `encode()` and receives a string in `__init__()` (to decode).

This class would have the method `level_up()`, `level_down_field`, and any other methods that make it easy to manipulate the URL but they work at the object level, and eventually you just call `encode()` and that's it.